packages/flutter_test/lib/src/accessibility.dart

```diff
@@ -438,80 +521,60 @@ class CustomMinimumContrastGuideline extends AccessibilityGuideline {
 class _ContrastReport {
-  /// Generates a contrast report given a list of colors.
-  ///
-  /// Given a list of integers [colors], each representing the color of a pixel
-  /// on a part of the screen, generates a contrast ratio report.
-  /// Each colors is given in ARGB format, as is the parameter for the
-  /// constructor [Color].
+  /// Generates a contrast report given color historgram.
   ///
   /// The contrast ratio of the most frequent light color and the most
   /// frequent dark color is calculated. Colors are divided into light and
   /// dark colors based on their lightness as an [HSLColor].
-  factory _ContrastReport(List<int> colors) {
-    final Map<int, int> colorHistogram = <int, int>{};
-    for (final int color in colors) {
-      colorHistogram[color] = (colorHistogram[color] ?? 0) + 1;
-    }
-    if (colorHistogram.length == 1) {
-      final Color hslColor = Color(colorHistogram.keys.first);
-      return _ContrastReport._(hslColor, hslColor);
-    }
-    // to determine the lighter and darker color, partition the colors
-    // by lightness and then choose the mode from each group.
-    double averageLightness = 0.0;
-    for (final int color in colorHistogram.keys) {
-      final HSLColor hslColor = HSLColor.fromColor(Color(color));
-      averageLightness += hslColor.lightness * colorHistogram[color]!;
+  ///
+  /// If text color is provided, it used instead of heuristicly guessed one.
+  factory _ContrastReport(Map<Color, int> colorHistogram, {Color? textColor}) {
+    // To determine the lighter and darker color, partition the colors
+    // by HSL lightness and then choose the mode from each group.
+    double totalLightness = 0.0;
+    int count = 0;
+    for (final MapEntry<Color, int> entry in colorHistogram.entries) {
+      totalLightness += HSLColor.fromColor(entry.key).lightness * entry.value;
+      count += entry.value;
     }
-    averageLightness /= colors.length;
+    final double averageLightness = totalLightness / count;
     assert(averageLightness != double.nan);
-    int lightColor = 0;
-    int darkColor = 0;
-    int lightCount = 0;
-    int darkCount = 0;
+
+    MapEntry<Color, int>? lightColor;
+    MapEntry<Color, int>? darkColor;
+
     // Find the most frequently occurring light and dark color.
-    for (final MapEntry<int, int> entry in colorHistogram.entries) {
-      final HSLColor color = HSLColor.fromColor(Color(entry.key));
+    for (final MapEntry<Color, int> entry in colorHistogram.entries) {
+      final double lightness = HSLColor.fromColor(entry.key).lightness;
       final int count = entry.value;
-      if (color.lightness <= averageLightness && count > darkCount) {
-        darkColor = entry.key;
-        darkCount = count;
-      } else if (color.lightness > averageLightness && count > lightCount) {
-        lightColor = entry.key;
-        lightCount = count;
+      if (lightness <= averageLightness) {
+        if (count > (darkColor?.value ?? 0)) {
```

Comment:
perhaps initialize `lightColor` and `darkColor` with 0 instead, making them non-nullable

Reply:
It is MapEntry. It is explicitly nullable to indicate if color is found.